vsintegration/src/FSharp.Editor/CommonHelpers.fs

```diff
@@ -0,0 +1,216 @@
+                        newData
+                    
+                lexState <- lineData.LexStateAtEndOfLine
+
+                if startLine <= i then
+                    result.AddRange(lineData.ClassifiedSpans |> Seq.filter(fun token ->
+                        textSpan.Contains(token.TextSpan.Start) ||
+                        textSpan.Contains(token.TextSpan.End - 1) ||
+                        (token.TextSpan.Start <= textSpan.Start && textSpan.End <= token.TextSpan.End)))
+
+            // If necessary, invalidate all subsequent lines after endLine
+            if endLine < lines.Count - 1 then 
+                match sourceTextData.[endLine+1] with 
+                | Some data  -> 
+                    if data.LexStateAtStartOfLine <> lexState then
+                        sourceTextData.ClearFrom (endLine+1)
+                | None -> ()
+
+            result
+        with ex -> 
+            Assert.Exception(ex)
+            reraise()  
+
+    let tryClassifyAtPosition (documentKey, sourceText: SourceText, filePath, defines, position: int, cancellationToken) =
+        let textLine = sourceText.Lines.GetLineFromPosition(position)
+        let textLinePos = sourceText.Lines.GetLinePosition(position)
+        let textLineColumn = textLinePos.Character
+
+        let classifiedSpanOption =
+            getColorizationData(documentKey, sourceText, textLine.Span, Some(filePath), defines, cancellationToken)
+            |> Seq.tryFind(fun classifiedSpan -> classifiedSpan.TextSpan.Contains(position))
+
+        match classifiedSpanOption with
+        | Some(classifiedSpan) ->
+            match classifiedSpan.ClassificationType with
+            | ClassificationTypeNames.ClassName
+            | ClassificationTypeNames.DelegateName
+            | ClassificationTypeNames.EnumName
+            | ClassificationTypeNames.InterfaceName
+            | ClassificationTypeNames.ModuleName
+            | ClassificationTypeNames.StructName
+            | ClassificationTypeNames.TypeParameterName
+            | ClassificationTypeNames.Identifier -> 
+                match QuickParse.GetCompleteIdentifierIsland true (textLine.ToString()) textLineColumn with
+                | Some (islandIdentifier, islandColumn, isQuoted) -> 
+                    let qualifiers = if isQuoted then [islandIdentifier] else islandIdentifier.Split '.' |> Array.toList
+                    Some (islandColumn, qualifiers, classifiedSpan.TextSpan)
+                | None -> None
+            | ClassificationTypeNames.Operator ->
+                let islandColumn = sourceText.Lines.GetLinePositionSpan(classifiedSpan.TextSpan).End.Character
+                Some (islandColumn, [""], classifiedSpan.TextSpan) 
+            | _ -> None
+        | _ -> None
+
+open System.Reflection
+open Microsoft.VisualStudio.Shell
+open System.Windows.Controls
+// A port of https://github.com/tomasr/viasfora/blob/master/Viasfora/Util/VsColors.cs
```

Comment:
1. We don't support _any_ VS version prior to 2017.
2. I've already removed that code.
3. We now use `IClassificationFormatMapService` https://github.com/Microsoft/visualfsharp/blob/master/vsintegration/src/FSharp.Editor/QuickInfoProvider.fs#L63